File: 240924/마법의 숲 탐색/magical-forest-exploration.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br;
	static StringBuilder sb;
	static StringTokenizer st;

	static final int[][] exitDirs = { { -1, 0 }, { 0, 1 }, { 1, 0 }, { 0, -1 } };

	static int R, C, K;
	static int[][] grid;
	static Set<Pos> exitSet;
	static int rowSum;

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
		sb = new StringBuilder();
		st = new StringTokenizer(br.readLine());
		R = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());
		exitSet = new HashSet<>();
		initializeGrid();
		rowSum = 0;
		int num = 2;
		for (int k = 0; k < K; k++) {
			st = new StringTokenizer(br.readLine());
			int cy = Integer.parseInt(st.nextToken());
			int d = Integer.parseInt(st.nextToken());
			int cx = 1;
			while (true) {
				if (grid[cx + 1][cy - 1] + grid[cx + 2][cy] + grid[cx + 1][cy + 1] == 0) {
					cx++;
				} else if (grid[cx - 1][cy - 1] + grid[cx][cy - 2] + grid[cx + 1][cy - 1] + grid[cx + 1][cy - 2]
						+ grid[cx + 2][cy - 1] == 0) {
					cx++;
					cy--;
					d = (d - 1) % 4;
				} else if (grid[cx + 1][cy + 1] + grid[cx][cy + 2] + grid[cx - 1][cy + 1] + grid[cx + 1][cy + 2]
						+ grid[cx + 2][cy + 1] == 0) {
					cx++;
					cy++;
					d = (d + 1) % 4;
				} else {
					break;
				}
			}
			if (cx < 4) {
				initializeGrid();
				exitSet.clear();
				continue;
			}
			grid[cx + 1][cy] = num;
			grid[cx - 1][cy] = num;
			for (int i = -1; i <= 1; i++) {
				grid[cx][cy + i] = num;
			}
			num++;
			d = (d + 4) % 4;
			exitSet.add(new Pos(cx + exitDirs[d][0], cy + exitDirs[d][1]));
			rowSum += bfs(cx, cy);
		}
		System.out.println(rowSum);
	}

	private static void initializeGrid() {
		grid = new int[R + 4][C + 2];
		for (int i = 0; i < R + 4; i++) {
			grid[i][0] = 1;
			grid[i][C + 1] = 1;
		}
		for (int i = 1; i < C + 1; i++) {
			grid[R + 3][i] = 1;
		}
	}

	private static int bfs(final int sx, final int sy) {
		final Queue<Pos> queue = new ArrayDeque<>();
		queue.add(new Pos(sx, sy));
		final boolean[][] visited = new boolean[R + 4][C + 2];
		visited[sx][sy] = true;
		int maxRow = 0;
		while (!queue.isEmpty()) {
			final Pos pos = queue.poll();
			if (maxRow < pos.x) {
				maxRow = pos.x;
			}
			for (int d = 0; d < 4; d++) {
				final int nx = pos.x + exitDirs[d][0];
				final int ny = pos.y + exitDirs[d][1];
				if (!visited[nx][ny] && (grid[pos.x][pos.y] == grid[nx][ny]
						|| (exitSet.contains(new Pos(pos.x, pos.y)) && grid[nx][ny] > 1))) {
					queue.add(new Pos(nx, ny));
					visited[nx][ny] = true;
				}
			}
		}
		return maxRow - 2;
	}

	static class Pos {

		final int x;
		final int y;

		public Pos(final int x, final int y) {
			this.x = x;
			this.y = y;
		}

		@Override
		public boolean equals(Object object) {
			final Pos other = (Pos) object;
			return this.x == other.x && this.y == other.y;
		}

		@Override
		public int hashCode() {
			return 100 * x + y;
		}
	}
}

File: 240924/마법의 숲 탐색/magical-forest-exploration.py
from collections import deque


def bfs(sx, sy):
    queue = deque([(sx, sy)])
    visited = [[False] * (C + 2) for _ in range(R + 4)]
    visited[sx][sy] = True
    max_row = 0
    while queue:
        x, y = queue.popleft()
        max_row = max(max_row, x)
        for dx, dy in exit_dirs:
            nx = x + dx
            ny = y + dy
            if not visited[nx][ny] and (grid[x][y] == grid[nx][ny] or ((x, y) in exit_set and grid[nx][ny] > 1)):
                queue.append((nx, ny))
                visited[nx][ny] = True
    return max_row - 2


R, C, K = map(int, input().split())
golem_infos = [tuple(map(int, input().split())) for _ in range(K)]
grid = [[1] + [0] * C + [1] for _ in range(R + 3)] + [[1] * (C + 2)]
exit_dirs = ((-1, 0), (0, 1), (1, 0), (0, -1))  # 북, 동, 남, 서
exit_set = set()
row_sum = 0
num = 2
for cy, d in golem_infos:
    cx = 1
    while True:
        if grid[cx + 1][cy - 1] + grid[cx + 2][cy] + grid[cx + 1][cy + 1] == 0:
            cx += 1
        elif grid[cx - 1][cy - 1] + grid[cx][cy - 2] + grid[cx + 1][cy - 1] \
                + grid[cx + 1][cy - 2] + grid[cx + 2][cy - 1] == 0:
            cx += 1
            cy -= 1
            d = (d - 1) % 4
        elif grid[cx + 1][cy + 1] + grid[cx][cy + 2] + grid[cx - 1][cy + 1] \
                + grid[cx + 1][cy + 2] + grid[cx + 2][cy + 1] == 0:
            cx += 1
            cy += 1
            d = (d + 1) % 4
        else:
            break
    if cx < 4:
        grid = [[1] + [0] * C + [1] for _ in range(R + 3)] + [[1] * (C + 2)]
        exit_set = set()
        continue
    grid[cx + 1][cy] = grid[cx - 1][cy] = num
    grid[cx][cy - 1:cy + 2] = [num] * 3
    num += 1
    exit_set.add((cx + exit_dirs[d][0], cy + exit_dirs[d][1]))
    row_sum += bfs(cx, cy)
print(row_sum)